File: node/test/groups/island.test.ts
/**
 * island.detail 测试：唯一的 island 分组端点，standard 鉴权方案，
 * x-qdmp-token-required: true（必须传 accessToken），query 参数为单个 id。
 */
import assert from 'node:assert/strict';
import {test} from 'node:test';
import type {MockAgent} from 'undici';

import {QdmpClient} from '../../src/client.js';
import {QdmpApiError} from '../../src/errors.js';
import {
  businessSuccess,
  createMockAgent,
  expectFailure,
  gatewayFailure,
} from '../helpers/mock-http.js';

function makeClient(mockAgent: MockAgent) {
  return new QdmpClient({
    appId: 'app-id-1',
    appSecret: 'app-secret-1',
    dispatcher: mockAgent,
  });
}

void test('island.detail: 正常成功路径，query.id 被正确透传，standard 请求头齐全', async () => {
  const {mockAgent, pool} = createMockAgent();
  let capturedHeaders: Record<string, string> | undefined;
  pool
    .intercept({
      path: '/island/v1/detail',
      method: 'GET',
      query: {id: 'island-1'},
      headers: h => {
        capturedHeaders = h as Record<string, string>;
        return true;
      },
    })
    .reply(
      200,
      businessSuccess({
        island: {id: 'island-1', name: 'Labubu 岛', joined: true},
      }),
    );

  const client = makeClient(mockAgent);
  const result = await client.island.detail(
    {accessToken: 'user-token-1'},
    {id: 'island-1'},
  );

  assert.equal(result.island?.id, 'island-1');
  assert.equal(result.island?.name, 'Labubu 岛');
  assert.equal(capturedHeaders?.['access-token'], 'user-token-1');
  assert.equal(capturedHeaders?.['x-echo-qdmp-version'], '1.0');
});

void test('island.detail: 缺失 accessToken 时本地直接报错，不发起任何 HTTP 请求', async () => {
  const {mockAgent, pool} = createMockAgent();
  pool
    .intercept({path: '/island/v1/detail', method: 'GET'})
    .reply(200, businessSuccess({island: {id: 'should-not-be-reached'}}));

  const client = makeClient(mockAgent);

  await expectFailure(() =>
    // @ts-expect-error 故意不传 accessToken，验证运行时本地校验
    client.island.detail({}, {id: 'island-1'}),
  );

  assert.equal(
    mockAgent.pendingInterceptors().length,
    1,
    '缺少 accessToken 时不应真的发出 HTTP 请求，拦截器必须仍处于未消费状态',
  );
});

void test('island.detail: 网关层错误信封 {code,message,details}（无 data/requestId）也必须被正确解析为失败', async () => {
  const {mockAgent, pool} = createMockAgent();
  pool
    .intercept({
      path: '/island/v1/detail',
      method: 'GET',
      query: {id: 'missing-island'},
    })
    .reply(404, gatewayFailure(5, 'island not found'));

  const client = makeClient(mockAgent);

  const err = await expectFailure(() =>
    client.island.detail({accessToken: 'user-token-1'}, {id: 'missing-island'}),
  );
  assert.ok(
    err instanceof QdmpApiError,
    `应解析为 QdmpApiError，实际抛出: ${err}`,
  );
  const apiErr = err as QdmpApiError;
  assert.equal(String(apiErr.code), '5');
  assert.equal(apiErr.httpStatus, 404);
});

File: go/island_test.go
package qdmp_test

// Contract exercised by this file:
//
//	asUser := client.WithAccessToken(token)
//	res, err := asUser.Island.Detail(ctx, generated.IslandDetailParams{Id: "..."})
//
// qdmp.IslandDetailResult{Island qdmp.IslandInfo}
// island.detail has x-qdmp-token-required=true (shared/generated/route-meta.json).

import (
	"context"
	"errors"
	"net/http"
	"testing"

	qdmp "github.com/EchoTechFE/qdmp-server-sdk/go"
	"github.com/EchoTechFE/qdmp-server-sdk/go/generated"
)

func TestIslandDetail_Success(t *testing.T) {
	const token = "island-detail-real-token"
	counter := newRequestCounter(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet || r.URL.Path != "/island/v1/detail" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if got := r.URL.Query().Get("id"); got != "island-1" {
			t.Fatalf("id query param = %q, want %q", got, "island-1")
		}
		if got := r.Header.Get("access-token"); got != token {
			t.Fatalf("access-token header = %q, want %q", got, token)
		}
		businessEnvelope(w, http.StatusOK, "0", "ok", "req-island-1", map[string]any{
			"island": map[string]any{
				"id":     "island-1",
				"name":   "哪吒岛",
				"image":  "https://example.com/island1.png",
				"joined": true,
			},
		})
	})
	srv := startServer(t, counter)
	client := newTestClient(t, srv.URL)

	res, err := client.WithAccessToken(token).Island.Detail(context.Background(), generated.IslandDetailParams{Id: "island-1"})
	if err != nil {
		t.Fatalf("Island.Detail() error = %v, want nil", err)
	}
	if res.Island.ID != "island-1" || res.Island.Name != "哪吒岛" || !res.Island.Joined {
		t.Fatalf("unexpected result: %+v", res)
	}
	if counter.Count() != 1 {
		t.Fatalf("server hit %d times, want 1", counter.Count())
	}
}

// TestIslandDetail_MissingAccessToken_NoRequestSent: island.detail is
// tokenRequired=true per shared/generated/route-meta.json, so an empty
// token must fail locally without ever reaching the network — even though
// this operation is confirmed to also work with an app-level token, the SDK
// still requires the caller to explicitly pass one.
func TestIslandDetail_MissingAccessToken_NoRequestSent(t *testing.T) {
	counter := newRequestCounter(func(w http.ResponseWriter, r *http.Request) {
		t.Fatalf("server should never have been called for a missing access token, got %s %s", r.Method, r.URL.Path)
	})
	srv := startServer(t, counter)
	client := newTestClient(t, srv.URL)

	res, err := client.WithAccessToken("").Island.Detail(context.Background(), generated.IslandDetailParams{Id: "island-1"})
	if err == nil {
		t.Fatalf("Island.Detail() error = nil, want a local validation error for empty accessToken")
	}
	if res != nil {
		t.Fatalf("Island.Detail() result = %+v, want nil on validation failure", res)
	}
	if !errors.Is(err, qdmp.ErrAccessTokenRequired) {
		t.Fatalf("err = %v, want errors.Is(err, qdmp.ErrAccessTokenRequired)", err)
	}
	if counter.Count() != 0 {
		t.Fatalf("server hit %d times, want 0", counter.Count())
	}
}

// TestIslandDetail_GatewayEnvelopeError exercises the gateway-style error
// envelope ({code, message, details}, no data/requestId) against island.detail
// too, not only the operation where it was first captured.
func TestIslandDetail_GatewayEnvelopeError(t *testing.T) {
	srv := startServer(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gatewayEnvelope(w, http.StatusInternalServerError, 13, "query parse error")
	}))
	client := newTestClient(t, srv.URL)

	res, err := client.WithAccessToken("some-token").Island.Detail(context.Background(), generated.IslandDetailParams{Id: "island-1"})
	if err == nil {
		t.Fatalf("Island.Detail() error = nil, want an error for the gateway-envelope failure shape")
	}
	if res != nil {
		t.Fatalf("Island.Detail() result = %+v, want nil on failure", res)
	}
	apiErr, ok := err.(*qdmp.QdmpApiError)
	if !ok {
		t.Fatalf("err type = %T, want *qdmp.QdmpApiError", err)
	}
	if apiErr.Code != "13" || apiErr.RequestID != "" {
		t.Fatalf("apiErr = %+v, want Code=13 and empty RequestID", apiErr)
	}
}
